### ISSUE TITLE
Master issue for Windows redraw performance woes

### ISSUE THREAD

**Issue by andlabs** [andlabs/libui#311]
Currently libui on Windows fakes transparency by drawing the parent control onto each child control as a background each time, resulting in awful performance. Not doing this with the current implementation of uiControl and control hierarchies results in controls indiscriminately having a default background color. I have two options: flatten the control hierarchy or split the work between WM_ERASEBKGND and WM_CTLCOLOR*. @lxn figured a lot of this out.

This issue will combine all existing issues on the matter.

This issue will be unlocked once all existing issues are migrated here.

**Comment by andlabs** [comment]
Huh, that was fewer issues than I thought there were...

**Comment by kdar** [comment]
I tried libui a year ago on win10 64bit and had flickering issues. Seems to still be the case.

![flicker](https://user-images.githubusercontent.com/119919/40882904-beb2a18e-66bd-11e8-9e92-207652371cbb.gif)

**Comment by andlabs** [comment]
Yes, that is what we are trying to resolve here. The question is how.

**Comment by msink** [comment]
So now - after 0.4 - most annoying is this issue.

Simetimes window opening takes 5 and more seconds, as in StapleButter/melonDS#220

**Comment by andlabs** [comment]
Yes, and addressing this issue is what's in store for Alpha 5's changes. I still need to figure out what that simpler solution lxn used was first, but there are several other advantages to just flattening the control hierarchy.

**Comment by msink** [comment]
If it should be almost fully rewritten anyway - maybe not invent a wheel and use [Yoga](https://yogalayout.com/) as layout engine?
Is it possible?

**Comment by tom-corwin** [comment]
@msink coming from a C# perspective, the C# bindings for yoga are very hacky, requiring (at least) 2 extra libraries on top of the actual Yoga library, plus a whole separate framework (Xamarin) for it to be used on macOS and Linux platforms. I think it would be more or a hinder than a help for alot of the bindings, though I can't speak for the others.

**Comment by msink** [comment]
Well - I mean not adding it as dependency, but include in libui as library.
Or at least implement similar interface and borrow some code.
Because there are WYSIWYG editors for it - would be great to reuse existing tools and infrastructure.

**Comment by tom-corwin** [comment]
> Well - I mean not adding it as dependency, but include in libui as library.

My apologies. I get your point now :)

> Or at least implement similar interface and borrow some code.
Because there are WYSIWYG editors for it - would be great to reuse existing tools and infrastructure.

I agree that it would be great to reuse existing tools and code to accomplish this.

**Comment by mleise** [comment]
I didn't realize performance could be an issue with a toolkit using native widgets. Best of luck with the solution! I see WM_PRINTCLIENT was already tried and @lxn seems confident that a more performant solution can be implemented.
